### PR TITLE
Add EngineSource, EngineManager, and Mayavi2 plugin panel

### PR DESCRIPTION
**Pull request by kitchoi**
Main features:
- [x] `EngineSource` for adapting CUDS from a simphony engine (analogous to `CUDSFileSource`) 
- [x] `EngineManager` for selecting one modeling engines out of many (TraitsUI)
- [x] `EngineManagerStandalone` for visualising datasets from engine, running and animating the results (GUI-free)
- [x] `EngineManagerStandaloneUI` a GUI version of `EngineManagerStandalone`
- [x] `EngineManagerMayavi2` as a panel to sit inside Mayavi2
- [x] `EngineManagerMayavi2Plugin` a envisage plugin for managing `EngineManagerMayavi2`
- [x] Added tests (except for `EngineManagerMayavi2`)
- [x] Added documentation
- [ ] Tests for `EngineManagerMayavi2`

Supporting elements:
- [x] `AddEngineSourceToMayavi`, `AddSourcePanel`, `RunAndAnimate` and `RunAndAnimatePanel` (for reusability)
- [x] `TabbedPanelCollection` collects multiple panels and displays as tabs
- [x] `default_modules` for setting default mayavi modules given the point_scalars_name/... (also see changes in `show.py`)

### CODE REVIEW

File: simphony_mayavi/modules/default_module.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    # default module for particle bonds
+    if (isinstance(source.cuds, ABCParticles) and
+            source.cuds.count_of(CUDSItem.BOND) > 0):
+        modules.append(default_bond_module())
+
+    if modules:
+        return modules
+    else:
+        message = "Unknown scalar/vector data for setting default module"
+        logging.warning(message)
```
Comment (dpinte) [inline simphony_mayavi/modules/default_module.py]: @kitchoi the typical way of using logging is to get a logger from the logging. If you don't target a specific logger, you can get one for this module:

```
logger = logging.getLogger(__name__)
```

then only use that logger instance for the calls:

```
logger.warning(xxx)
```

File: simphony_mayavi/plugins/add_source_panel.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                     selected="_pending_source",
+                     selected_index="_pending_source_index")),
+            ),
+        title="Visualize")
+
+    # --------------------------------------------------
+    # Public methods
+    # --------------------------------------------------
+
+    def __init__(self, engine_name, engine, mayavi_engine):
```
Comment (dpinte) [inline simphony_mayavi/plugins/add_source_panel.py]: you typically avoid doing this with `HasTraits` class. In this case, you break the traits initialization and event system. 

You can remove the `__init__` method and use the standard trait intialization by keyword:

```
AddSourcePanel(engine=engine, engine_name=engine_name, mayavi_engine=mayavi_engine)
```

If you really, really need an `__init__`  method, you should calls `HasTraits.__init__(self, **kwargs)` to properly intialize the `HasTraits` class (with the appropriate definitions of the `kwargs`)

File: simphony_mayavi/plugins/add_source_panel.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        for visualization
+    '''
+    engine = Instance(ABCModelingEngine)
+    engine_name = Str
+    mayavi_engine = Instance("mayavi.core.engine.Engine")
+
+    label = "Add to Mayavi"
+
+    # Buttons for the UI
+    _add_dataset = Button("+")
```
Comment (dpinte) [inline simphony_mayavi/plugins/add_source_panel.py]: I tend to prefer using a `Event` trait with a `ButtonEditor` in the UI

File: simphony_mayavi/plugins/engine_manager.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        Simphony Modeling Engine
+    '''
+    # Engines that are added to the manager
+    engines = Dict(Str, Instance(ABCModelingEngine))
+
+    # Names of engines in the Manager
+    _engine_names = ListStr
+
+    # Selected engine (overloaded BasicPanel)
+    engine = Property
```
Comment (dpinte) [inline simphony_mayavi/plugins/engine_manager.py]: for the property, you tycally pass the type of the property and what it depends on

File: simphony_mayavi/plugins/engine_manager_mayavi2.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    VIEWS = "envisage.ui.workbench.views"
+
+    service_offers = List(contributes_to=SERVICE_OFFERS)
+    views = List(contributes_to=VIEWS)
+
+    # Private methods.
+    def _service_offers_default(self):
+        """ Trait initializer. """
+        worker_service_offer = ServiceOffer(
+            protocol=('simphony_mayavi.plugins.engine_manager_mayavi2.'
```
Comment (dpinte) [inline simphony_mayavi/plugins/engine_manager_mayavi2.py]: it's better to keep the protocol name as one single string (even if too long) and add a `# noqa`. If not, searching for that string will always fail on those entries.

File: simphony_mayavi/plugins/engine_manager_standalone_ui.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    panels = Instance(TabbedPanelCollection)
+
+    traits_view = View(
+        VGroup(
+            Group(Item("engine_name", label="Engine Wrapper")),
+            Group(Item("panels", style="custom", show_label=False))),
+        title="Engine Manager")
+
+    def __init__(self, engine_name, engine, mayavi_engine=None):
```
Comment (dpinte) [inline simphony_mayavi/plugins/engine_manager_standalone_ui.py]: if `EngineManager` is a `HasTraits` class, this `__init__` is broken

File: simphony_mayavi/plugins/run_and_animate_panel.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+    def _number_of_time_steps_changed(self):
+        value = self.number_of_time_steps
+        self.engine.CM[CUBA.NUMBER_OF_TIME_STEPS] = value
+
+    # ----------------------------------------------------------
+    # Public methods
+    # ----------------------------------------------------------
+
+    def __init__(self, engine, mayavi_engine):
```
Comment (dpinte) [inline simphony_mayavi/plugins/run_and_animate_panel.py]: `HasTraits.__init__`

File: simphony_mayavi/plugins/tests/test_add_engine_source_to_mayavi.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+from mayavi.core.api import NullEngine
+
+from simphony_mayavi.plugins.add_engine_source_to_mayavi import (
+    add_source_and_modules_to_scene, AddEngineSourceToMayavi)
+from simphony_mayavi.sources.api import EngineSource
+from simphony_mayavi.sources.tests import testing_utils
+
+
+class TestAddEngineSourceToMayavi(unittest.TestCase):
```
Comment (dpinte) [inline simphony_mayavi/plugins/tests/test_add_engine_source_to_mayavi.py:11]: We should talk about testing traits class during our call today

**Comment by dpinte** [comment]
@kitchoi great work! See my comments in the code

**Comment by kitchoi** [comment]
@dpinte Thanks for reviewing the code!
I have:
- added trait's tests (`traits.testing.UnittestTools`)  for `EngineManager`, `AddSourcePanel` and `EngineManagerStandaloneUI` (`AddEngineSourceToMayavi` and `RunAndAnimate` are not subclasses of `HasTrait`)
- kept the `__init__` for `EngineManagerStandaloneUI` for it to be consistent with `EngineManagerStandalone`
- `__init__` in `AddSourcePanel`, `RunAndAnimatePanel` and `TabbedPanelCollection` are removed.
- replaced `logging.warn/info/..` with `logger.warn/info/...`
- added more tests for `default_module`

Forgot to ask, is it possible to test `EngineManagerMayavi2` and `EngineManagerMayavi2Plugin` as it is being loaded by envisage?

**Comment by kitchoi** [comment]
@dpinte is this good to merge?

As for loading engine wrappers from *.py file or from default factory function, I am planning to put this into a separate PR:
add_engine_panel

**Comment by dpinte** [comment]
@kitchoi LGTM. And a separate PR for the latest additions is definitely a good idea

**Comment by kitchoi** [comment]
Thanks!

**Pull request merged**